File: java/src/main/java/sa/amnn/kit/AmenApiError.java
package sa.amnn.kit;

import java.util.List;

/** Any non-2xx response. {@code codes} holds the API's error codes, e.g. "price__required". */
public class AmenApiError extends RuntimeException {
    public final int status; public final List<String> codes; public final String method, path, body;
    public AmenApiError(int status, List<String> codes, String method, String path, String body) {
        super(status + " " + method + " " + path + ": " + (codes.isEmpty() ? body : String.join(", ", codes)));
        this.status = status; this.codes = codes; this.method = method; this.path = path; this.body = body;
    }
    public boolean has(String code) { return codes.contains(code); }
    public boolean isRetryable() { return status == 429 || status >= 500; }
}

File: php/src/Amen/AmenApiError.php
<?php
declare(strict_types=1);
namespace Amen;

/** Any non-2xx response. `$codes` holds the API's error codes, e.g. "price__required". */
class AmenApiError extends \RuntimeException
{
    /** @param string[] $codes */
    public function __construct(public readonly int $status, public readonly array $codes, public readonly string $method,
                                public readonly string $path, public readonly mixed $body = null)
    {
        parent::__construct("$status $method $path: " . (implode(', ', $codes) ?: json_encode($body)), $status);
    }
    public function has(string $code): bool { return in_array($code, $this->codes, true); }
    public function isRetryable(): bool { return $this->status === 429 || $this->status >= 500; }
}

/** Thrown locally, before any HTTP call, when an action is not valid for the deal's status. */
class AmenLifecycleError extends \LogicException {}
